src/transform/src/table_elimination.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software matches_collection governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Detects an input being unioned with its negation and cancels them out
+
+use crate::{TransformArgs, TransformError};
+use expr::MirRelationExpr;
+
+/// Detects an input being unioned with its negation and cancels them out
+#[derive(Debug)]
+pub struct TableElimination;
+
+impl crate::Transform for TableElimination {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), TransformError> {
+        relation.try_visit_mut(&mut |e| self.action(e))
+    }
+}
+
+impl TableElimination {
+    /// Detects an input being unioned with its negation and cancels them out
+    pub fn action(&self, relation: &mut MirRelationExpr) -> Result<(), TransformError> {
+        if let MirRelationExpr::Union { base, inputs } = relation {
+            let matching_negation = |input: &MirRelationExpr,
+                                     inputs: &[MirRelationExpr],
+                                     start_idx: usize|
+             -> Option<usize> {
+                match input {
+                    MirRelationExpr::Negate { input: inner_input } => {
+                        for i in start_idx..inputs.len() {
+                            if inputs[i] == **inner_input {
+                                return Some(i);
+                            }
+                        }
+                    }
+                    _ => {
+                        for i in start_idx..inputs.len() {
+                            if let MirRelationExpr::Negate { input: inner_input } = &inputs[i] {
+                                if *input == **inner_input {
+                                    return Some(i);
+                                }
+                            }
+                        }
+                    }
+                }
+                None
+            };
+
+            if let Some(j) = matching_negation(&*base, inputs, 0) {
+                Self::cancel_relation(&mut *base);
+                Self::cancel_relation(&mut inputs[j]);
+            }
+
+            for i in 0..inputs.len() {
+                if let Some(j) = matching_negation(&inputs[i], inputs, i + 1) {
+                    Self::cancel_relation(&mut inputs[i]);
+                    Self::cancel_relation(&mut inputs[j]);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn cancel_relation(relation: &mut MirRelationExpr) {
+        *relation = MirRelationExpr::constant(vec![], relation.typ());
```

Comment:
Hmm...given that the helper method is one line long, perhaps it's better to just replace the references to `Self::cancel_relation` with just the `input = MirRelationExpr::constant(...)`. 

By the way, since 
1) MirRelationExpr::typ is considered to be "not especially cheap" and "should be used judiciously", 
https://github.com/MaterializeInc/materialize/blob/7470ce56a6beae78d47f295fc03bec287ac6b07d/src/expr/src/relation/mod.rs#L231
2) All the inputs in the `Union` have the same column types (not necessarily the same keys, but the keys don't matter anymore when you're canceling things out).

I suggest halving the number of `MirRelationExpr::typ` calls by:
Step 1: define `let relation_typ = input1.typ()` once when you're about to cancel out two inputs.
Step 2: Replace the first input with `MirRelationExpr::constant(vec![], relation_typ.clone())`
Step 3: Replace the second input with `MirRelationExpr::constant(vec![], relation_typ)`

An alternative is to precompute `relation_typ` at the beginning of this method, but I think that would result in unnecessary work given that most queries will not be in a situation where canceling out two inputs is possible.

Reply:
Done as suggested. Thanks!